File: src/impl/softfloat/f16.rs
use fpa_specr_berkeley_softfloat::*;

use crate::RoundingMode;

#[inline]
fn from_f16(value: f16) -> float16_t {
    float16_t { v: value.to_bits() }
}

#[inline]
fn to_f16(value: float16_t) -> f16 {
    f16::from_bits(value.v)
}

/// Returns `a + b` as specific rounding mode.
#[inline]
pub fn round_add(a: f16, b: f16, mode: &RoundingMode) -> f16 {
    let r = unsafe { c_add_f16(mode.as_berkeley_softfloat(), from_f16(a), from_f16(b)) };
    to_f16(r)
}

/// Returns `a - b` as specific rounding mode.
#[inline]
pub fn round_sub(a: f16, b: f16, mode: &RoundingMode) -> f16 {
    let r = unsafe { c_sub_f16(mode.as_berkeley_softfloat(), from_f16(a), from_f16(b)) };
    to_f16(r)
}

/// Returns `a * b` as specific rounding mode.
#[inline]
pub fn round_mul(a: f16, b: f16, mode: &RoundingMode) -> f16 {
    let r = unsafe { c_mul_f16(mode.as_berkeley_softfloat(), from_f16(a), from_f16(b)) };
    to_f16(r)
}

/// Returns `a * b` as specific rounding mode.
#[inline]
pub fn round_div(a: f16, b: f16, mode: &RoundingMode) -> f16 {
    let r = unsafe { c_div_f16(mode.as_berkeley_softfloat(), from_f16(a), from_f16(b)) };
    to_f16(r)
}

/// Returns `a * b + c` with single rounding (fused multiply-add) as specific rounding mode.
#[inline]
pub fn round_mul_add(a: f16, b: f16, c: f16, mode: &RoundingMode) -> f16 {
    let r = unsafe {
        c_fma_f16(
            mode.as_berkeley_softfloat(),
            from_f16(a),
            from_f16(b),
            from_f16(c),
        )
    };
    to_f16(r)
}

/// Returns `a.sqrt()` as specific rounding mode.
#[inline]
pub fn round_sqrt(a: f16, mode: &RoundingMode) -> f16 {
    let r = unsafe { c_sqrt_f16(mode.as_berkeley_softfloat(), from_f16(a)) };
    to_f16(r)
}

File: srmfpa-berkeley-softfloat-3/src/c/f16.c
#include "softfloat.h"

float16_t c_add_f16(uint_fast8_t const mode, float16_t const a, float16_t const b) {
  volatile uint_fast8_t const dflt = softfloat_roundingMode;
  softfloat_roundingMode = mode;

  volatile float16_t const temp = f16_add(a, b);

  softfloat_roundingMode = dflt;

  return temp;
}

float16_t c_sub_f16(uint_fast8_t const mode, float16_t const a, float16_t const b) {
  volatile uint_fast8_t const dflt = softfloat_roundingMode;
  softfloat_roundingMode = mode;

  volatile float16_t const temp = f16_sub(a, b);

  softfloat_roundingMode = dflt;

  return temp;
}

float16_t c_mul_f16(uint_fast8_t const mode, float16_t const a, float16_t const b) {
  volatile uint_fast8_t const dflt = softfloat_roundingMode;
  softfloat_roundingMode = mode;

  volatile float16_t const temp = f16_mul(a, b);

  softfloat_roundingMode = dflt;

  return temp;
}

float16_t c_div_f16(uint_fast8_t const mode, float16_t const a, float16_t const b) {
  volatile uint_fast8_t const dflt = softfloat_roundingMode;
  softfloat_roundingMode = mode;

  volatile float16_t const temp = f16_div(a, b);

  softfloat_roundingMode = dflt;

  return temp;
}

float16_t c_fma_f16(uint_fast8_t const mode, float16_t const a, float16_t const b, float16_t const c) {
  volatile uint_fast8_t const dflt = softfloat_roundingMode;
  softfloat_roundingMode = mode;

  volatile float16_t const temp = f16_mulAdd(a, b, c);

  softfloat_roundingMode = dflt;

  return temp;
}

float16_t c_sqrt_f16(uint_fast8_t const mode, float16_t const a) {
  volatile uint_fast8_t const dflt = softfloat_roundingMode;
  softfloat_roundingMode = mode;

  volatile float16_t const temp = f16_sqrt(a);

  softfloat_roundingMode = dflt;

  return temp;
}
